src/Dialect/ONNX/ONNXOps/Tensor/Reshape.cpp

```diff
@@ -116,10 +118,12 @@ LogicalResult ONNXReshapeOp::verify() {
 
 LogicalResult ONNXReshapeOp::inferShapes(
     std::function<void(Region &)> doShapeInference) {
-  // Cannot infer shape without data rank or static shape of shape.
+  // Cannot infer shape without data rank and static shape of shape.
+  // If shape is constant shape, the rank of the output known.
+  // This step may be helpful to reach the fix point.
   // TODO: Infer shape without data rank if shape is a constant
   //       without -1 and without 0 and allowzero.
-  if (!hasShapeAndRank(getData()) || !hasStaticShape(getShape().getType()))
+  if (!hasShapeAndRank(getData()) && !hasStaticShape(getShape().getType()))
```

Comment:
As the comment at L125 said, even though `shape` of `Reshape` is a constant, it may contain symbolic value, e.g. `0`, meaning taking the dimension from the input dimension. Not sure the code afterward can handle such a case: input is unranked and shape contains `0`. Otherwise, we need to check `shape` values in details.

Reply:
Add the code to handle 0 or -1 for shape. Since we just need to handle them for shape inference, those dimensions are conservatively given unknown (QuestionmarkIndexExpr). When coming to lowering, the data can not be unranked.

Reply:
Also test cases are added to lit test.

Reply:
Thanks!